### PR TITLE
test: Enable SC2046 and SC2086 shellcheck rules

### LINKED ISSUE #20879: lint-shell: Enable SC2046 and SC2086, prevent unintentional word-splitting
We currently disable the `shellcheck` rules SC2046 and SC2086 in our `lint-shell.sh` script, however, I think it is important that we enable them and make our scripts conformant and therefore more robust.

I have encountered many inscrutable build failures where the ultimate culprit was missing quoting, and from personal experience, I think word-splitting is the exception rather than the norm, and it would make our scripts easier to read if instances of word-splitting are pointed out by the shellcheck directives.

### PR DESCRIPTION
**Pull request by hebasto**
Closes #20879.
Replaces #22695.

**Note for reviewers**. Some touched shell scripts are not being run in CI, therefore they require more thorough reviewing:
- `contrib/devtools/gen-manpages.sh`
- `contrib/macdeploy/detached-sig-apply.sh`
- `contrib/windeploy/detached-sig-create.sh`
- `src/qt/res/animation/makespinner.sh`

### CODE REVIEW

**Comment by dougEfresh** [comment]
Since you are already touching the ci scripts with double quotes  `"` 
Maybe add `"` around the  `cd P_CI_DIR` in  DOCKER_EXEC function: 

```
DOCKER_CI_CMD_PREFIX bash -c "export PATH=$BASE_SCRATCH_DIR/bins/:\$PATH && cd \"$P_CI_DIR\" && $*"
```

**Comment by hebasto** [comment]
Friendly ping @dongcarl and @practicalswift 
:tiger2:

**Comment by Zero-1729** [comment]
Approach ACK

File: ci/test/00_setup_env_native_qt5.sh [hunk 1]
```diff
@@ -15,5 +15,5 @@ export RUN_UNIT_TESTS_SEQUENTIAL="true"
 export RUN_UNIT_TESTS="false"
 export GOAL="install"
 export PREVIOUS_RELEASES_TO_DOWNLOAD="v0.15.2 v0.16.3 v0.17.2 v0.18.1 v0.19.1 v0.20.1"
-export BITCOIN_CONFIG="--enable-zmq --with-libs=no --with-gui=qt5 --enable-reduce-exports
+export BITCOIN_CONFIG="--enable-zmq --with-libs=no --with-gui=qt5 --enable-reduce-exports \
```
Comment (MarcoFalke) [inline ci/test/00_setup_env_native_qt5.sh:18]: Does this change change behavior?

Comment (hebasto) [reply ci/test/00_setup_env_native_qt5.sh:18]: No, it doesn't. Why do you think so?

Comment (MarcoFalke) [reply ci/test/00_setup_env_native_qt5.sh:18]: Ok, was just wondering if there is a difference.

Comment (hebasto) [reply ci/test/00_setup_env_native_qt5.sh:18]: More details.

Being quoted, the `"$BITCOIN_CONFIG` preserves the `newline` character.

Now it is [treated](https://www.gnu.org/software/bash/manual/html_node/Escape-Character.html#Escape-Character) as a line continuation.

**Comment by hebasto** [comment]
Updated 02e9579cecc8002e37013dc11a0b7ab5b7a90d42 -> 27fbb10cbc5fb82e52e5b62dc34cd5c5c729452e ([pr23462.03](https://github.com/hebasto/bitcoin/commits/pr23462.03) -> [pr23462.04](https://github.com/hebasto/bitcoin/commits/pr23462.04)):

- rebased due to the conflict with #23458
- addressed @dougEfresh's [comment](https://github.com/bitcoin/bitcoin/pull/23462#issuecomment-962671897)

**Comment by laanwj** [comment]
Tested successfully:
- `contrib/devtools/gen-manpages.sh`
- `src/qt/res/animation/makespinner.sh`

**Comment by ifuckingyouyeah** [comment]
download
kiss my dick bitch

**Review by Zero-1729 — COMMENTED** [review]
tACK 27fbb10cbc5fb82e52e5b62dc34cd5c5c729452e

Tested on macOS v12.0.1 (21A559)

Successfully tested the following files locally and can verify the SC2046 & SC2086 warnings fixed:

> **Note**: On macOS, the default `sed` triggers an error when running `gen-manpages.sh`. This is resolved by using the GNU version `gnu-sed` instead (i.e. install and prepend its path in `PATH` to override the default `sed`). Running `src/qt/res/animation/makespinner.sh` also requires `imagemagick`. `contrib/macdeploy/detached-sig-apply.sh` requires [signapple](https://github.com/achow101/signapple) installed.

- contrib/devtools/gen-manpages.sh
- src/qt/res/animation/makespinner.sh
- contrib/macdeploy/detached-sig-apply.sh

**PS**: the minor notes above are intended only for those who haven't run these files before on macOS.

**Comment by laanwj** [comment]
> Note: On macOS, the default sed triggers an error when running gen-manpages.sh.

Wrt general gen-manpages.sh frustrations: please see #22681. The sed parts are likely going to be removed completely before 0.23.

File: ci/test/04_install.sh [hunk 2]
```diff
@@ -12,6 +12,7 @@ fi
 
 if [ "$CI_OS_NAME" == "macos" ]; then
   sudo -H pip3 install --upgrade pip
+  # shellcheck disable=SC2086
```
Comment (laanwj) [inline ci/test/04_install.sh:15]: What is the idea of disabling these specific checks in some places? Is the idea that these are temporary and need to be fixed later? (but too much hassle to fix in thir PS?)

Comment (hebasto) [reply ci/test/04_install.sh:15]: > What is the idea of disabling these specific checks in some places?

```bash
$ PIP_PACKAGES="zmq lief"
$ pip3 install --user "$PIP_PACKAGES"
ERROR: Invalid requirement: 'zmq lief'
```

Comment (hebasto) [reply ci/test/04_install.sh:15]: Looks like that. And that code requires `# shellcheck disable=SC2086`.

Comment (hebasto) [reply ci/test/04_install.sh:15]: Also see https://github.com/koalaman/shellcheck/wiki/SC2086#exceptions

Comment (hebasto) [reply ci/test/04_install.sh:15]: This works:
```bash
$ PIP_PACKAGES=(zmq lief)
$ pip3 install --user "${PIP_PACKAGES[@]}"
```

**Comment by dongcarl** [comment]
Concept ACK, unintentional globbing and word splitting lead to inscrutable bugs.

We should mark with a `disable=` wherever we need them or simply use arrays.

File: ci/test/04_install.sh (hunk 2 above)
Comment (laanwj) [reply ci/test/04_install.sh:15]: Using arrays sounds like a good idea to me. At least for scripts like this that explicitly need `bash`.

**Comment by hebasto** [comment]
> Using arrays sounds like a good idea to me. At least for scripts like this that explicitly need `bash`.

> We should mark with a `disable=` wherever we need them or simply use arrays.

Could we leave introducing arrays for a follow up?

**Comment by laanwj** [comment]
> Could we leave introducing arrays for a follow up?

Fine with me. Merging this soon closes off new such issues being introduced. And it's already the second try.

**Comment by hebasto** [comment]
Updated 27fbb10cbc5fb82e52e5b62dc34cd5c5c729452e -> fe0ff569ea6c353f88609c0f5f9b6fa80ff74f15 ([pr23462.04](https://github.com/hebasto/bitcoin/commits/pr23462.04) -> [pr23462.05](https://github.com/hebasto/bitcoin/commits/pr23462.05)):

- rebased due to the conflict with #23114
- the second commit has been re-worked, and discovered bugs have been fixed

> What is the idea of disabling these specific checks in some places?

Now this PR introduces `# shellcheck disable=SC2046` in two places only:
- `contrib/windeploy/detached-sig-create.sh` is not a `bash` script, that is required to use arrays
- in `test/lint/lint-python.sh` too much hassle is required to avoid it

**Comment by laanwj** [comment]
Code review re-ACK fe0ff569ea6c353f88609c0f5f9b6fa80ff74f15

**Comment by fanquake** [comment]
Possible this broke the verify commits pre push hook. See https://github.com/bitcoin-core/bitcoin-maintainer-tools/pull/198 / https://github.com/bitcoin/bitcoin/pull/35235.

**Pull request merged**